### PR TITLE
New approach for handling keyboard layouts and keyboard events

### LINKED ISSUE #1401: Support Monaco key chords properly
Afaik Theia now supports key sequences, so I think it could be possible to register monaco keybindings of types `Chord`. Related code is [here](https://github.com/theia-ide/theia/blob/master/packages/monaco/src/browser/monaco-keybinding.ts#L43).

### LINKED ISSUE #3885: "Alt Gr" key not working properly on Windows
The `AltRight` key behaves differently on Windows than on other OSs: It is reported both as `AltRight` and `ControlLeft` by the browser (see e.g. https://github.com/electron/electron/issues/13895). That key is very important on non-US keyboards to access characters such as `[`, `]`, `{`, `}` etc.

The change #3833 is very problematic in this context. For example, on a French keyboard the keystroke `AltGr + 5` should produce a `[` character, but now it maps to `Ctrl + [` instead, which means _decrease indentation_ in Monaco.

### LINKED ISSUE #4079: discussion: [keybinding] Properly manage keybindings and layouts
Following a bunch of issues regarding keybindings, and trying to understand what VS Code does, it is clear that this is not an easy topic.

## What Theia does

From what I understand, at the implementation level, we check properties in the following order:
1. [`KeyboardEvent.keyCode`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode) (number) *(deprecated)*
2. [`KeyboardEvent.keyIdentifier`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyIdentifier) (string) *(deprecated)*
3. [`KeyboardEvent.which`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/which) (number) *(deprecated)*
4. and finally [`KeyboardEvent.code`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code) (string) which is the sole non-deprecated property that we check, but we only do it as last resort :thinking:.

From that lookup, we check against a big map of `KEY_CODE_TO_KEY` (number to Key) and `CODE_TO_KEY` (string to Key) in order to get our own `Key` abstraction, defining the right `code` and `keyCode` for said key (see this as a the bridge between some `event.code` and its `US` representation as a `event.keyCode`).

## Proposal

Wouldn't be too bad, except that maybe we should forget about `keyCode`, and instead start using [`KeyboardEvent.key`](https://developer.mozilla.org/fr/docs/Web/API/KeyboardEvent/key).

We can see how this works by playing on the following website:
https://w3c.github.io/uievents/tools/key-event-viewer.html

From my tests, doing the following key presses:

keyboard-us-de

- (#&#x2060;1 to #&#x2060;5) `ctrl+/` (US)
- (#&#x2060;6 to #&#x2060;10) `ctrl+-` (DE) physical equivalent of `ctrl+/` (US)
- (#&#x2060;11 to #&#x2060;17) `ctrl+shift+7` (DE) same localized representation as `ctrl+/` (US)
  (on a german keyboard, one must do `shift+7` to output a `/`)

What we can see is that the `event.key` field always match the originally defined binding, when trying to get a localized `ctrl+/`.

## `code` and `key` against `keyCode`

`code` has the advantage of being *locale independent*, while `key` is *locale dependent*.

The inconvenient of having a `key` that is local dependent is that we can precompute a `CODE_TO_KEY`, but *cannot* do it for a `LOCALIZED_KEY_TO_KEY` map because the localized `key` field will vary from one layout to the other.

Other difference between `KeyboardEvent.key` and `KeyboardEvent.keyCode` is that the value of `key` will be affected by the modifiers:

- press `7` (US): `event.key === '7'`
- press `shift+7` (US): `event.key === '&'`

Using a German keyboard:

- press `7` (DE): `event.key === '7'`
- press `shift+7` (DE): `event.key === '/'`

This means that it makes us dependent on the way we refer to keys: Do we want to say that you need to type on a key where `7` is drawn (somehow) and hold `shift`, no matter what `shift+7` resolves to???
(`/` on a German layout or `&` on an US layout)

In our case, **we don't actually refer to keys this way**: when we define `ctrl+/` on a US layout to comment a line in the editor, the representational keybinding **stays the same** on the German keyboard, it is just that if we were to represent the physical keys that would be involved back on a US keyboard, it would **look like** we did `ctrl+shift+7` (when actually it would be `ctrl+&` once we simplify the binding).

All in all, if we kept looking at `event.key`, we wouldn't have had any problem between US and DE.

## Keyboard layouts

This brings us to the management of keyboard layouts: there is currently no API in browsers to query some user's keyboard layout.

So what I propose is to have the user explicitly choose its keyboard layout from Theia's UI, with a dropdown or anything, and have keyboard layouts contributed somehow (via `ContributionProvider`s and/or plugins). This means that these contributions will need to be explicitly written down somehow.

We can store the chosen layout in the browser's local storage in some global field, and ask the user for it on the very first launch of Theia if it is undefined (because first run). Or we can default to US, but still display a notification to guide the user on how to change it.

This might allow us then to display the correct information when displaying bindings, based on what layout was chosen.

## Contributing/Writing keyboard layouts

Contributing layouts and creating them would be like creating TextMate themes (but simplier). Just do it once, and everyone is happy. Even if it doesn't map 100% to the real deal, 75% would be better than nothing at all.

**[Our current implementation](https://github.com/theia-ide/theia/blob/master/packages/core/src/browser/keys.ts#L555-L883) already defines something similar to a US layout map**. Although a bit sketchy in my opinion. It did the work until now, but maybe since we are having problems today it would be a good idea to clean that up?

From the top of my head, a keyboard layout map would look like the following, taking into account `event.code` and `event.key`:

```ts
const US_LOCALIZED_KEY_TO_PHYSICAL_KEYS_RAW = {
    'a': { code: 'KeyA' },
    'A': { code: 'KeyA', shift: true },
    '/': { code: 'Slash' },
    '?' { code: 'Slash', shift: true },
    ...
};
```

From there we could use this data structure to compute forward and reverse maps for any lookup.

To be fair, having to map each localized character to the corresponding physical key + modifiers pains me, but that's the only way to be able to correctly encode and decode from one layout to the other.

Also again: we are already partially doing this.

(_Worst case scenario_ we only do it for the US keyboard, to hell the rest?)

The design could be improved, this is just an idea that I throw out there.

# More testing

Maybe there's even too much information, let's take a look at different keyboards, for the same keybinding: `ctrl+/` to toggle line comment.

- `ctrl+/` (US) base for comparison
- `ctrl+/` (DE) maps to `ctrl+&` (US) or `ctrl+shift+7` (US)
- `ctrl+/` (FR) maps to `ctrl+>` (US) or `ctrl+shift+.` (US)

**Please note that these are exactly the keys I need to press on my US keyboard when I use the foreign layouts on VS Code. Remember that you need to close and reopen the whole VS Code window between layout switches for it to have an effect.**

We actually might not care what the localized keybinding maps to on an US keyboard. The issue might be that we currently do the translation, when we shouldn't even do it (maybe).

So what do we get from mapping localized bindings to their complete physical sequence? If we were to map some bindings to physical keys (instead of localized keys) then it would matter, in order to actually display the right information on screen when changing layout (if the feature gets added).

On the other hand, it would also allow us to understand alternative notations, such as `ctrl+&` could be written `ctrl+shift+7`. Not sure if it is something we want to support, but it is out there.

Example:

- `shift+alt(right)+4` (US) that maps to `£` (US, International with dead keys - IWDK).

In this case, the lookup of the `event.key = £` would fail when using US (IWDK), because we only have the standard US layout defined (default). But on failure, we could still try a lookup based on the physical keys, in that case `{ code: 'Digit4', shift: true, altRight: true }`.

image

- (#&#x2060;1 to #&#x2060;14) US layout
- (#&#x2060;15 to #&#x2060;29) US (IWDK) layout

This case is super weird, because it seems that we should use the function `getModifierState(code)` in order to have a better range on the available modifiers, as on my keyboard I do have something called `AltGraph` which is mapped to the physical `AltRight` key. And there seems to be a difference between `AltLeft` and `AltRight`, but not on US standard.

I wouldn't know what to do in that case, but this is why I write it down, maybe you have a better idea :)

We should definitely take a look at https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key/Key_Values for the available values for `event.code` and `event.getModifierState(code)`.

## What VS Code does

VS Code -by running on the same host the frontend/backend- have access to node APIs, and they use their own [`native-keymap`](https://www.npmjs.com/package/native-keymap) native npm module.

Although their internals are filled with different big data maps/enums ([1](https://github.com/Microsoft/vscode/blob/master/src/vs/base/common/keyCodes.ts#L8-L199), [2](https://github.com/Microsoft/vscode/blob/master/src/vs/base/common/keyCodes.ts#L229-L359), [3](https://github.com/Microsoft/vscode/blob/master/src/vs/base/common/keyCodes.ts#L8-L199), [4](https://github.com/Microsoft/vscode/blob/master/src/vs/base/common/keyCodes.ts#L229-L359), [5](https://github.com/Microsoft/vscode/blob/master/src/vs/base/common/scanCode.ts#L11-L208), [6](https://github.com/Microsoft/vscode/blob/master/src/vs/base/common/scanCode.ts#L268-L467)) and their relations are not quite easy to follow... I wouldn't know if we should go through the pain of inspiring ourselves from their implementation, seeing how intricate it is.

Something that I noticed with their design though, is that in VS Code the layout seems to be fetched on window opening, and even if you change the layout at runtime it will not dynamically update.

It is even worse than that, because even if you change the layout while running, the keybindings stay bound to the old physical layout that was set (doing `ctrl+/` (US) after starting with a DE layout still trigger the `decrease zoom` as if we pressed `ctrl+-` (DE) which involves the same physical keys... this looks like a bug to me).

## Foot notes

This doesn't account for cross OS layout, as I don't know if there is a difference between a Linux US layout and a Mac US layout...

I wrote this document in order to share what I think I could understand about this keybinding issues, the current system drove me crazy, and I don't think it has to be this complex. I also believe it came at this point because people kept trying to "fix" it without having the big picture.

I do not claim to have understood it all, if anything this *must* be seen as a resource for anyone wanting to fix the keybindings in Theia. Please add comments and any insight you could have on the matter. I really rely on you all there to review this and eventually come with better fixes, as I won't be able to focus on this task more than just thinking about it at high level like I just did, sorry :(

I think keybindings are important as I use them around 8 hours a day 7/7. So I would expect Theia to be good in that area, yet it is not really the case.

## Interesting references:

- Documentation about `KeyboardEvent.code`: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key/Key_Values
- Keyboard event tester: https://w3c.github.io/uievents/tools/key-event-viewer.html
- List of deprecated and valid properties for `KeyboardEvent`: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent

### PR DESCRIPTION
**Pull request by spoenemann**
Closes #4079, closes #1401, closes #3885, maybe also #2200 and #3518 (and possibly more).

I refactored a lot of the existing keybindings related code and introduced two new services:

 * KeyboardLayoutProvider (`layout-provider.ts`) tries to get the user's keyboard layout in the format defined by the [native-keymap](https://www.npmjs.com/package/native-keymap) package.
 * KeyboardLayoutService (`keyboard-layout-service.ts`) maps keybindings according to the provided keyboard layout and determines the characters to display in the UI (menus, quick open etc.)

The KeyboardLayoutProvider behaves differently depending on the platform and browser:

 * On Electron I use the `native-keymap` package in the backend and transfer that information to the frontend via JSON-RPC.
 * On browsers I load static JSON files that have been prepared with the script `generate-layout.js`, which simply writes the output of `native-keymap` to a file. I have already created layout files for German, French and US layouts.
 * On Chrome there is partial support for the experimental [Keyboard API](https://developer.mozilla.org/en-US/docs/Web/API/Keyboard). I find the user's layout by comparing all keys in the Keyboard API mapping with the statically loaded layouts.
 * On browsers with no support for the Keyboard API, I use `navigator.language` to guess the keyboard by the user's language setting (this is highly error-prone, of course).

The KeyboardLayoutService behaves differently depending on the platform:

 * On Mac OS and Linux I compute a reverse mapping from reachable characters to the respective KeyCodes, i.e. keys with possible `shift` and `alt` modifiers. This reverse mapping is then used to transform keybindings. For example, the keybinding `cmd+/` requires the `/` character, which is reachable via `shift+7` on German Mac keyboards. The reverse mapping transforms the keybinding to `cmd+shift+7`.
 * On Windows every key has a so-called _virtual key_ which is provided by `native-keymap`. Here the reverse mapping is computed based on the virtual keys. For example, the virtual key identifier for `/` is `VK_OEM_2`. On German Mac keyboards, this virtual key is mapped to the key labeled `#`, so the keybinding `ctrl+/` is transformed to `ctrl+#`.

The platform-specific behavior should be equivalent to what VS Code does. Please tell me if you find keybindings that are still different in VS Code.

There is still a lot of work to be done in order to improve keyboard detection for browsers that don't support the Keyboard API. For example, we could implement some heuristics that detect the keyboard after the user has pressed some keys, or implement a UI for selecting a keyboard. Furthermore, we should add more static layouts. Should we extract the script to a separate repository to make that easier also for external contributors? We don't need to build the whole of Theia just to run a simple NodeJS script.

The `native-keymap` package has a method to receive notifications when the keyboard layout has changed in the system. Unfortunately, this does not work yet (https://github.com/Microsoft/node-native-keymap/issues/10).

I have created a CQ for the new dependency to `native-keymap`:
https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19423
I also had to copy the typings from VS Code (`native-keymap.d.ts`) because they are not included in the package. Do we need another CQ for this?

### CODE REVIEW

File: packages/core/src/browser/keys.ts [hunk 1]
```diff
@@ -1,900 +0,0 @@
-/********************************************************************************
```
Comment (akosyakov) [inline packages/core/src/browser/keys.ts]: Could we keep it here to avoid breaking changes?

**Comment by benoitf** [comment]
hello, is it planned to be included in the upcoming release ?

File: packages/core/src/browser/keys.ts (hunk 1 above)
Comment (spoenemann) [reply packages/core/src/browser/keys.ts]: As far as I could tell that code is not used a lot outside of `core`. IMO moving it to the new subfolder helps to keep the code organized.

Comment (akosyakov) [reply packages/core/src/browser/keys.ts]: I cannot tell anything about extensions and products outside theia repo. For them we’ve agree to go through deprecation policy when it’s possible.

**Comment by marechal-p** [comment]
Detail, but the build on Linux now requires additional libraries, see:
https://travis-ci.org/theia-ide/theia/jobs/511526642#L788

It is also not possible to build this branch on Gitpod because of this.

File: packages/core/src/browser/keyboard/keyboard-browser.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+    protected get allLayouts(): NativeKeyboardLayout[] {
+        return [
+            this.winUS, this.macUS, this.winFrench, this.macFrench, this.winGerman, this.macGerman
+        ];
+    }
+
+    private client?: KeyboardLayoutClient;
+
+    setClient(client?: KeyboardLayoutClient): void {
```
Comment (marechal-p) [inline packages/core/src/browser/keyboard/keyboard-browser.ts]: Why use this pattern over some event emitter able to handle multiple clients?

In this case there can only be one client for this provider (which is in singleton scope).

Comment (marechal-p) [reply packages/core/src/browser/keyboard/keyboard-browser.ts]: Nevermind, just understood we had to interact with `KeyboardLayoutService` and not this class directly, my bad :)

File: packages/core/src/browser/keys.ts (hunk 1 above)
Comment (akosyakov) [reply packages/core/src/browser/keys.ts]: We actually rely on such imports to install listeners for the base widgets, so i would expect that custom widgets are importing them, as tree and mini browser widgets do in Theia repo.

You could do something like:
```ts
import { KeyCode, KeyModifier, ... } from './keyboard';
export { KeyCode, KeyModifier };
```
If you want to restructure without breaking.

**Comment by akosyakov** [comment]
> hello, is it planned to be included in the upcoming release ?

Maybe we better to test it in next for a month before putting in latest? That's a cool change, but we have to verify that there are not surprises. cc @marcdumais-work @svenefftinge

**Comment by spoenemann** [comment]
> the build on Linux now requires additional libraries, see:
> https://travis-ci.org/theia-ide/theia/jobs/511526642#L788

Yes, we need the additional package `libxkbfile-dev`:

https://github.com/Microsoft/node-native-keymap#installing

I added it to our `.travis.yml`.

> It is also not possible to build this branch on Gitpod because of this.

@svenefftinge should we add a Dockerfile to this repository, or would you rather add `libxkbfile-dev` to workspace-full?

File: packages/core/src/browser/keyboard/keyboard-layout-service.ts [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+                    key: Key.getKey(code),
+                    character: value
+                });
+            }
+        }
+    }
+
+    private getCharacterIndex(key: Key, shift?: boolean): number {
+        if (shift) {
+            return 256 + key.keyCode;
```
[the remark this answers is not in the document]
Comment (spoenemann) [reply packages/core/src/browser/keyboard/keyboard-layout-service.ts]: Replaced with a constant

File: packages/core/src/browser/keybinding.ts [hunk 4]
```diff
@@ -81,6 +76,10 @@ export interface Keybinding {
     when?: string;
 }
 
+export interface ResolvedKeybinding extends Keybinding {
+    resolved?: KeyCode[];
```
[the remark this answers is not in the document]
Comment (spoenemann) [reply packages/core/src/browser/keybinding.ts]: I'm definitely pro well-commented code! I'll add some comments where they make sense.

**Comment by spoenemann** [comment]
I restructured some of the code according to https://github.com/Microsoft/node-native-keymap/issues/10#issuecomment-477233420

**Comment by akosyakov** [comment]
@spoenemann could you rebase please, i can start reviewing afterwards

**Comment by spoenemann** [comment]
Done.

File: packages/core/src/node/backend-application-module.ts [hunk 5]
```diff
@@ -90,4 +92,12 @@ export const backendApplicationModule = new ContainerModule(bind => {
...
         return new ApplicationPackage({ projectPath });
     }).inSingletonScope();
+
+    bind(NativeKeyboardLayoutProvider).toSelf().inSingletonScope();
+    bind(KeyboardLayoutProvider).toService(NativeKeyboardLayoutProvider);
+    bind(ConnectionHandler).toDynamicValue(ctx =>
+        new JsonRpcConnectionHandler(keyboardPath, () =>
+            ctx.container.get(KeyboardLayoutProvider)
+        )
+    );
```
Comment (akosyakov) [inline packages/core/src/node/backend-application-module.ts]: `.inSingletonScope();`

File: packages/monaco/src/typings/monaco/index.d.ts [hunk 6]
```diff
@@ -403,25 +403,90 @@ declare module monaco.keybindings {
...
+        public abstract getDispatchParts(): [string | null, string | null];
         /**
          * Returns the firstPart, chordPart of the keybinding.
          * For simple keybindings, the second element will be null.
          */
-        public abstract getParts(): [ResolvedKeybindingPart, ResolvedKeybindingPart | undefined];
+        public abstract getParts(): [ResolvedKeybindingPart | null, ResolvedKeybindingPart | null];
     }
 
     export class USLayoutResolvedKeybinding extends ResolvedKeybinding {
```
Comment (akosyakov) [inline packages/monaco/src/typings/monaco/index.d.ts]: Is it used? or can be removed?

File: packages/terminal/src/browser/terminal-frontend-contribution.ts [hunk 7]
```diff
@@ -225,7 +225,8 @@ export class TerminalFrontendContribution implements TerminalService, CommandCon
         });
         keybindings.registerKeybinding({
             command: TerminalCommands.TERMINAL_CLEAR.id,
-            keybinding: 'ctrlcmd+k'
+            keybinding: 'ctrlcmd+k',
+            context: TerminalKeybindingContexts.terminalActive
```
Comment (akosyakov) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts]: why is it not necessary anymore? was it a bug?

**Comment by akosyakov** [comment]
@spoenemann i've rebased the PR and push one commit to enable Electron testing in Gitpod, please pull before force pushing next time

File: packages/core/src/common/keyboard/layout-provider.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+export const KeyboardLayoutProvider = Symbol('KeyboardLayoutProvider');
+
+export interface KeyboardLayoutProvider {
+    getNativeLayout(): Promise<NativeKeyboardLayout>;
+}
+
+export const KeyboardLayoutChangeNotifier = Symbol('KeyboardLayoutChangeNotifier');
+
+export interface KeyboardLayoutChangeNotifier {
+    onNativeLayoutChanged: Event<NativeKeyboardLayout>;
```
Comment (akosyakov) [inline packages/core/src/common/keyboard/layout-provider.ts]: Is it fired then a user change keyboard layout? I'm testing changing from EN to DE and back, and it does not work for me, i had to reload Chrome. (on mac)

File: packages/core/src/browser/keyboard/keyboard-browser.ts [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, postConstruct } from 'inversify';
+import { isOSX, isWindows } from '../../common/os';
+import { Emitter } from '../../common/event';
+import { NativeKeyboardLayout, KeyboardLayoutProvider, KeyboardLayoutChangeNotifier } from '../../common/keyboard/layout-provider';
+
+@injectable()
+export class BrowserKeyboardLayoutProvider implements KeyboardLayoutProvider, KeyboardLayoutChangeNotifier {
```
Comment (akosyakov) [inline packages/core/src/browser/keyboard/keyboard-browser.ts]: please name files after classes `browser-keyboard-layout-provider` to simplify a search by a class name in Theia and dev tools, it's in our naming conventions: https://github.com/theia-ide/theia/wiki/Coding-Guidelines

File: packages/core/src/common/keyboard/layout-provider.ts (hunk 8 above)
Comment (akosyakov) [reply packages/core/src/common/keyboard/layout-provider.ts]: looks like `layoutchange` event is not implemented on mac :(

File: packages/core/src/node/keyboard/native-layout.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import * as nativeKeymap from 'native-keymap';
+import { injectable } from 'inversify';
+import { KeyboardLayoutProvider, NativeKeyboardLayout } from '../../common/keyboard/layout-provider';
+
+@injectable()
+export class NativeKeyboardLayoutProvider implements KeyboardLayoutProvider {
```
Comment (akosyakov) [inline packages/core/src/node/keyboard/native-layout.ts]: please give proper name to all files, `native-layout` does not reflect class name and that it has any rel to keyboard, following the guidline and naming files after classes is the best

**Comment by akosyakov** [comment]
That's a great improvement! Shortcuts work way better now any existing web IDE i've tried!

Testing Electron part... 

@spoenemann would be fine if I do one pass and fix naming of files/events to follow ours naming conventions?

File: packages/core/src/common/keyboard/layout-provider.ts (hunk 8 above)
Comment (marechal-p) [reply packages/core/src/common/keyboard/layout-provider.ts]: I am also not seeing layout updates at runtime when running Electron on Linux.

But the event doesn't seem to be implemented on Linux at all, so it seems normal:
https://github.com/Microsoft/node-native-keymap/blob/master/src/keyboard_x.cc#L235-L237

**Comment by marechal-p** [comment]
@akosyakov to be fair, I don't see much more else to comment about.

When your review comments will be addressed I will be entirely good with these changes!

Even if we happen to see something later, the code looks better now so I expect issues to be trivial.

File: packages/monaco/src/typings/monaco/index.d.ts (hunk 6 above)
Comment (akosyakov) [reply packages/monaco/src/typings/monaco/index.d.ts]: nvm found there it's used

**Comment by akosyakov** [comment]
@marechal-p as far as I know @spoenemann is going to open 2 more issues and work on them:
- quick pick as you suggested if the layout can b auto-recognized
- perf optimizations to move detection to the backend

But if the layout is not recognized then it works as on master for your. Not unexpected behaviour?

**Comment by marechal-p** [comment]
@akosyakov I only noticed that using the German keyboard, it is not possible to do `ctrl+/` in order to toggle line comments, using `[LeftCtrl] + [LeftShift] + [Digit7]` or `[LeftCtrl] + [Slash]`.

But it feels like expected behavior, as that will be fixed by being able to select the current layout.

**Comment by akosyakov** [comment]
>  I only noticed that using the German keyboard, it is not possible to do ctrl+/ in order to toggle line comments, using [LeftCtrl] + [LeftShift] + [Digit7] or [LeftCtrl] + [Slash].

@marechal-p the same on master, so it's fine

**Comment by marechal-p** [comment]
Everything is green on my side then :)

Amazing work by the way, I really like these changes!

**Comment by akosyakov** [comment]
kudos to @spoenemann :)

File: packages/core/src/common/keyboard/layout-provider.ts (hunk 8 above)
Comment (spoenemann) [reply packages/core/src/common/keyboard/layout-provider.ts]: https://github.com/Microsoft/vscode/issues/17521#issuecomment-267933654

> we now detect keyboard layout changes on OSX, no reasonable solution was found on Linux.

**Comment by spoenemann** [comment]
What about my question:

> I have created a CQ for the new dependency to native-keymap:
> https://dev.eclipse.org/ipzilla/show_bug.cgi?id=19423
> I also had to copy the typings from VS Code (native-keymap.d.ts) because they are not included in the package. Do we need another CQ for this?

**Comment by akosyakov** [comment]
@spoenemann yes, we need a CQ for all copied code

**Comment by spoenemann** [comment]
Is it a "Contribution of code to be maintained by an Eclipse Foundation project" or a "Third-Party Code Request"?

**Comment by marcdumais-work** [comment]
Hi @spoenemann ,

The CQ looks good, thanks.

**Pull request merged**